media-proxy/src/session-rdma-rx.cc

```diff
@@ -0,0 +1,165 @@
+    if (err) {
+        ERROR("%s ep_recv_buf failed with errno: %s", __func__, fi_strerror(-err));
+        return err;
+    }
+    return 0;
+}
+
+void RxRdmaSession::handle_received_buffers()
+{
+    shm_buf_info_t *buf_info;
+    int err;
+    uint16_t bursted_buf_num;
+
+    err = ep_cq_read(ep_ctx, (void **)&buf_info, 1);
+    if (err) {
+        if (err != -EAGAIN)
+            INFO("%s ep_rxcq_read: %s", __func__, strerror(-err));
+        return;
+    }
+    fb_recv++;
+
+    err = memif_tx_burst(memif_conn, 0, &buf_info->shm_buf, 1, &bursted_buf_num);
+    if (err != MEMIF_ERR_SUCCESS && bursted_buf_num != 1) {
+        INFO("%s memif_tx_burst: %s", __func__, memif_strerror(err));
+        return;
+    }
+    buf_info->used = false;
+}
+
+void RxRdmaSession::frame_thread()
+{
+    while (!atomic_load_explicit(&shm_ready, std::memory_order_acquire) && !stop)
+        usleep(1000);
+
+    INFO("%s, RX RDMA thread started\n", __func__);
+    while (!stop) {
+        if (!atomic_load_explicit(&shm_ready, std::memory_order_acquire))
+            continue;
+        while (!pass_empty_buf_to_libfabric()) {
+        }
+        handle_received_buffers();
+    }
+}
+
+RxRdmaSession::RxRdmaSession(libfabric_ctx *dev_handle, const mcm_conn_param &request,
+                             memif_ops_t &memif_ops)
+    : Session(memif_ops, request.payload_type, RX), ep_cfg{0}, ep_ctx(0), stop(false),
+      frame_thread_handle(0), fb_recv(0), shm_bufs(0)
+{
+    transfer_size = request.payload_args.rdma_args.transfer_size;
+
+    ep_cfg.rdma_ctx = dev_handle;
+    memcpy(&ep_cfg.remote_addr, &request.remote_addr, sizeof(request.remote_addr));
+    memcpy(&ep_cfg.local_addr, &request.local_addr, sizeof(request.local_addr));
+    ep_cfg.dir = direction::RX;
+}
+
+int RxRdmaSession::init()
+{
+    int err = ep_init(&ep_ctx, &ep_cfg);
+    if (err) {
+        ERROR("Failed to initialize libfabric's end point");
+        return -1;
+    }
+
+    shm_buf_num = 1 << 4;
+    shm_bufs = (shm_buf_info_t *)calloc(shm_buf_num, sizeof(shm_buf_info_t));
+    if (!shm_bufs) {
+        ERROR("Failed to allocate memory");
+        return -1;
+    }
+
+    err = shm_init(transfer_size, 4);
+    if (err < 0) {
+        ERROR("Failed to initialize shared memory");
+        return -1;
+    }
+
+    frame_thread_handle = new std::thread(&RxRdmaSession::frame_thread, this);
+    if (!frame_thread_handle) {
+        ERROR("Failed to create thread");
+        return -1;
+    }
+    return 0;
+}
+
+RxRdmaSession::~RxRdmaSession()
+{
+    INFO("%s, fb_recv %d\n", __func__, fb_recv);
+    stop = true;
+    if (ep_ctx) {
+        ep_destroy(&ep_ctx);
+        ep_ctx = 0;
+    }
+
+    if (shm_bufs) {
+        free(shm_bufs);
+        shm_bufs = 0;
+    }
+
+    if (frame_thread_handle) {
+        frame_thread_handle->join();
+        delete frame_thread_handle;
+        frame_thread_handle = 0;
+    }
+}
+
+int RxRdmaSession::on_connect_cb(memif_conn_handle_t conn)
```

Comment:
Please add empty method on_disconnect_cb in RxRdma and Tx RDMA and write inside the comment:
/* TODO: unregister in libfabric memory regions allocated by memif */
Thanks.

Now it is not an issue as we disconnect only when session is destroyed, but in case when we would like to disconnect and connect different memif connection to existing session it will be a problem.

Reply:
fix in https://github.com/OpenVisualCloud/Media-Communications-Mesh/pull/240/commits/3d694d9aae84c488290fdbf5558135eea4dc2b21